File: TwoSum_day_9/twosum.py
# Brute force method
# // Time complexity O(n^2)
# // space complexity O(n)

def twoSum(arr,target):
    for i in range (len(arr)):
        for j in range (i+1 ,len(arr)):
            if arr[i] + arr[j]  == target:
               return[arr[i],arr[j]]
    return []



# optimized way 

# // Time complexity O(n)
# // space complexity O(n) 

def twoSumNum(arr,target) :

    seen = set()

    for num in arr:
        needed = seen - target

        if needed in seen:
                return[needed,num]

        seen.add(num)

    return[num]


# Sorting

def twoSumSorting(arr,target):

    left = 0
    right = len(arr) -1

    while(left <right):
        sum = arr[left] +arr[right]

        if sum == target:
            return [arr[left],arr[right]]

        if(sum > target):
            right -+ 1
        else:
            left += 1

    return[]

File: TwoSum_day_9/twoSum.js
// Brute force method 
// Time complexity O(n^2)
// space complexity O(n)

function twoSum(arr, target){

    for(let i=0; i<arr.length; i++){
        for(let j = i+1; j<arr.length; j++){
            if(arr[i] + arr[j] === target){
                return [arr[i] , arr[j]]
            }
        }
    }

    return []

}


// optimized way 

// Time complexity O(n)
// space complexity O(n)

function checkTwoSumNum(arr,target){

    const seen = new Set()

    for(const num of arr){
        const needed = seen - target

        if(seen.has(needed)){
            return [needed,num]
        }
        seen.add(num)
    }
    return []
}


// Sorting method 
// Time: O(n log n)


function twoSumSort(arr,target){
 let left =0
 let right = arr.length-1

 while(left<right){
    const sum = arr[left] + arr[right]

    if(sum === target){
        return [arr[left],arr[right]]
    }

    if(sum >target){
        right --;
    }else{
        left++
    }
 }

 return []
}